Add aria-label to code and HTML block textarea elements. #2027

tfrommen opened the pull request:
## Description
Implemented the suggestion in #2027: add missing textarea labels as `aria-label` attributes.

## How Has This Been Tested?
Only visually. I don't have a screen reader, nor am I familiar with software in that regard.

## Types of changes
* Add `aria-label` attribute to code and HTML block textarea elements.
* Add `placeholder` attribute to HTML block.

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

tfrommen committed:
Add aria-label to code and HTML block textarea elements. #2027
Also, add placeholder to HTML block.

afercia commented:
@tfrommen thanks! Rather than describing the block type, the labels should indicate what the texareas are about or what usera are supposed to enter. Worth noting all the block containers are already labeled to indicate the block type, so this info would be basically duplicated:

![screen shot 2017-10-05 at 00 32 26](https://user-images.githubusercontent.com/1682452/31202921-0099dc5a-a965-11e7-9031-5981c323cd83.png)

tfrommen commented:
Ah, OK. Makes sense.

I guess I misinterpreted the "Youtube URL" in your example a little bit. I thought the label should describe what it is, and then the placeholder what to do with it. But I'll adapt that.

Do you have any suggestions for the exact wording? I mean, duplicating the `placeholder` also is not good, right?

afercia commented:
> I mean, duplicating the placeholder also is not good, right?

Yep but in this case the placeholder usage is not correct. Placeholders should be used exclusively for a short [hint about the sample value or the expected format](https://www.w3.org/TR/html51/sec-forms.html#the-placeholder-attribute) but seems we're having some difficulties in communicating this to designers 🙂  

I'd use labels as we'd normally do in a form: is it a "name" field? The label would be "Name". Is it a custom HTML textarea? I'd simply use "HTML".

tfrommen commented:
@afercia quoting what you wrote in #2027:

> Worth noting other input fields, for example the embed URL input, use both an aria-label and a placeholder attribute, e.g.:
> `aria-label="YouTube URL" placeholder="Enter URL to embed here…"`
> 
> This pattern seems better to me, and some consistency across all these kind of fields would be nice.

I thought I followed that scheme when using, for example, "Code block" and "Enter code...".
Now, I totally get that blocks already have a dedicated label, so it's redundant to state something like "Code block" here.

So, should I make the new `aria-label`s have the current value of the placeholders, and remove the placeholders?

afercia commented:
Personally, I'd remove the placeholders because they're used incorrectly and against the spec. There's also a Trac ticket about this, see https://core.trac.wordpress.org/ticket/40331

Designers tend to use placeholders as visual replacements for labels and that's not correct. I guess they won't agree on the removal though, so this would need a discussion to get to a consensus.

tfrommen committed:
Update aria-label, and remove (superfluous) placeholder. #2027

tfrommen commented:
@afercia I now removed the label (i.e., the one that was there before, and also the one I introduced for consistency), and adapted the new labels' values.

Are you good with this?

afercia commented:
@tfrommen looks good from an a11y perspective 👍

tfrommen commented:
Any non-a11y opinions on this simple PR?

youknowriad commented:
Added a design review label. Personally, I think placeholders are important for novice users even if this is not their "correct" usage. But I'd leave the decision to our designers.

youknowriad committed:
Code Block: Restore the placeholder

youknowriad commented:
Rebased and updated this old PR to fix only the missing labels. Let's leave the placeholders as is.
